### PR TITLE
Support implicit alignment

### PR DESCRIPTION
**Pull request by SUPERCILEX**
Fixes https://github.com/typst/typst/issues/695

### CODE REVIEW

**Comment by laurmaedje** [comment]
Doesn't this reduce expressiveness? Now I can't align something across two lines while ignoring the third one? For example, from your test where the third line can be ignored right now.
```
$
"abc" &= c \
&= c + 1 & "By definition" \
&= d + 100 + 1000 \
&= x && "Even longer" \
$
```
Maybe this behaviour is rarely desirable, let me think some more about this. Would also be cool to integrate this kind of alignment capability into normal non-math layouts and want to think a bit about the consequences of that.

**Comment by SUPERCILEX** [comment]
> Maybe this behaviour is rarely desirable.

Yeah, the current behavior treats & as an alignment between other &s whereas I've always thought of & as creating an implicit table and delimiting the columns. If there's ever a desire for the old behavior, I think we could add a flag to the equation function.

**Comment by Enivex** [comment]
The main problem with alignment now isn't the lack of implicit &, but the fact that alignment doesn't alternate between right and left. That makes many things hard (or impossible) to express.

That being said, I wouldn't mind if implicit & was an option, because it's very rare that you want text to appear as far left as in the given example.

**Comment by SUPERCILEX** [comment]
So it sounds like this needs 3 features:
- Option to disable implicit columns.
- Option to control spacing between alignments.
- Flipping left/right bias after each alignment.

Not sure where that leaves this PR... I think this could go through as long as I add an option to go back to the previous behavior. Thoughts? Then we can fix the other stuff.

**Comment by laurmaedje** [comment]
I think I'm against an option. We should just decide, otherwise it's just confusing. Flipping between left/and right seems like a thing we should clearly do (I would be happy about a PR for that!). Regarding this PR: I had never thought about LaTeX's & as a table (even though its also used in actual tables!). In that thinking, your proposal makes sense. Maybe let's first get the left/right thing and then revisit this?

**Comment by SUPERCILEX** [comment]
Sounds good. I'll take a look at flipping rn. In the meantime, I created https://github.com/typst/typst/pull/891 to extract out the single-pass algorithm.

File: library/src/math/align.rs [hunk 1]
```diff
@@ -28,21 +28,31 @@ pub(super) fn alignments(rows: &[MathRow]) -> AlignmentResult {
     for row in rows {
         let mut width = Abs::zero();
         let mut alignment_index = 0;
-        for fragment in row.iter() {
-            if matches!(fragment, MathFragment::Align) {
+
+        macro_rules! update_width {
```
Comment (laurmaedje) [inline library/src/math/align.rs:32]: could you try to write this without a macro?

**Review by laurmaedje — COMMENTED** [review]
alright, I guess it makes sense to follow what LaTeX does here

File: library/src/math/align.rs (hunk 1 above)
Comment (SUPERCILEX) [reply library/src/math/align.rs]: Leads to code duplication which is a bummer, but sounds good.

**Pull request merged**